lib/src/inline_syntaxes/footnote_ref_syntax.dart

```diff
@@ -0,0 +1,32 @@
+import '../ast.dart';
+import '../charcode.dart';
+import '../inline_parser.dart';
+import 'inline_syntax.dart';
+
+class FootnoteRefSyntax extends InlineSyntax {
+  FootnoteRefSyntax() : super(r'!?\[\^(\S+?)\]');
+
+  @override
+  bool onMatch(InlineParser parser, Match match) {
+    final label = match[1]!;
+    final refs = parser.document.footnoteReferences;
+    if (!refs.containsKey(label)) {
```

Comment:
Since `footnoteReferences` is typed with non-nullable types, we can simplify some of this code by grabbing the value first with `final something = parser.document.footnoteReferences[label];`. Then we can guard this if-statement with `if (something != null)` and line 19 can be `parser.document.footnoteReferences[label] = something + 1;`